In [ ]:
Using PyHandle
==============

Training material for ...  in Barcelona.

Sofiane Bendoukha, May 2017

What is PyHandle?
-----------------

> The PyHandle Python library is a client library for interaction with a Handle System server, using the native REST interface introduced in Handle System 8. The library offers methods to create, update and delete Handles as well as advanced functionality such as searching over Handles using an additional search servlet and managing multiple location entries per Handle.

> It also allows to interact with database hosting the handle records (out of scope of this training).

> In this training, we will the use the PyHandle pre-release.


Preparation for this training
-------------------------

**Requirements**

The library currently supports Python 2.6, Python 2.7 and Python 3.5 and requires access to a Handle System server (at 
least version 8.1). The library requires OpenSSL v1.0.1 or higher.

**Getting pyhandle**

* > You can get PyHandle from GITHUB (soon on PyPi).
  > git clone https://github.com/EUDAT-B2SAFE/PYHANDLE.git


**Install your own Jupyter Notebook (and pyhandle) **

Linux:

~~~
# Download conda
wget https://repo.continuum.io/miniconda/Miniconda2-latest-Linux-x86_64.sh

# Run the installer
bash /home/.../Downloads/Miniconda2-latest-Linux-x86_64.sh

# Create a virtual environment and activate it:
se

source activate rda_may2016

# Install jupyter:
conda install jupyter

# Install pyhandle:
cd ~/rda_hh_2016/library/pyhandle-0.0.1rc0/
python setup.py install

# Run the notebook:
jupyter notebook
~~~

Other systems:
On windows, the steps are very similar. For the detailed instructions, please see the slides.

In [ ]:
Using PyHandle
==============

Training material for ...  in Barcelona.

Sofiane Bendoukha, May 2017

What is PyHandle?
-----------------

> The PyHandle Python library is a client library for interaction with a Handle System server, using the native REST interface introduced in Handle System 8. The library offers methods to create, update and delete Handles as well as advanced functionality such as searching over Handles using an additional search servlet and managing multiple location entries per Handle.

> It also allows to interact with database hosting the handle records (out of scope of this training).

> In this training, we will the use the PyHandle pre-release.


Preparation for this training
-------------------------

**Requirements**

The library currently supports Python 2.6, Python 2.7 and Python 3.5 and requires access to a Handle System server (at 
least version 8.1). The library requires OpenSSL v1.0.1 or higher.

**Getting pyhandle**

* > You can get PyHandle from GITHUB (soon on PyPi).
  * git clone https://github.com/EUDAT-B2SAFE/PYHANDLE.git
* cd PYHANDLE



**Install your own Jupyter Notebook (and pyhandle) **

Linux:

~~~
# Download conda
wget https://repo.continuum.io/miniconda/Miniconda2-latest-Linux-x86_64.sh

# Run the installer
bash /home/.../Downloads/Miniconda2-latest-Linux-x86_64.sh

# Create a virtual environment and activate it:
se

source activate rda_may2016

# Install jupyter:
conda install jupyter

# Install pyhandle:
cd ~/rda_hh_2016/library/pyhandle-0.0.1rc0/
python setup.py install

# Run the notebook:
jupyter notebook
~~~

Other systems:
On windows, the steps are very similar. For the detailed instructions, please see the slides.

(1) Import
==========

First, we have to import the PyHandle library. The library is used by creating a client object and using its methods to interact with the Global Handle System.

The object that we are going to interact with is called the `Client`, so let's import it directly.

In [1]:
#from b2handle.handleclient import HandleClient as Client

(If you use B2Handle 1.0.x, use this import statement instead:)

In [3]:
from b2handle.handleclient import EUDATHandleClient as Client

The help() method gives us useful information about its methods.

In [4]:
# help(Client) # Attention, this gives a long result!

> Task 1a: Import pyhandle from the provided egg.

> Task 1b: Use the help() function to find out which methods the Client provides.


(2) Resolving handles
=================

It is easy to resolve a handle and read its handle record using the Pyhandle library. For this, we instantiate the client in read-mode and use its reading methods.

(2.1) Instantiation of the client
---------------------------

For using the client in read-only mode, no further information is necessary. It works out-of-the-box:

Now we can use its various reading methods, for example `get_value_from_handle(handle)` or `retrieve_handle_record(handle)`.

For example, `retrieve_handle_record(handle)` returns a dictionary of the record's entries:

In [5]:
handle = '21.T14998/TEST'
record = client.retrieve_handle_record(handle)
print type(record)

<type 'dict'>


In [6]:
print record

{u'URL': 'http://www.dkrz.de', u'VERSION_NUMBER': '1', u'HS_ADMIN': "{u'index': 200, u'handle': u'0.NA/21.T14998', u'permissions': u'011111110011'}", u'IS_PART_OF': 'blablabla'}


But it's nicer to pretty-print the record:

In [8]:
#pretty_record = client.pretty_print_handle_record(handle)
#print pretty_record

In [10]:
value1 = client.get_value_from_handle(handle, 'URL')
value2 = client.get_value_from_handle(handle, 'IS_PART_OF')
print value1
print value2

http://www.dkrz.de
blablabla


> Task 2.1a: Instantiate the pyhandle client for read-only access.

> Task 2.1b: Retrieve entire record of handle *10876.test/rdapidxyz* as a dictionary.

> Task 2.1c: Retrieve a single handle value from the record.

> Task 2.1d: Pretty-print the handle record.

> Task 2.1e: [advanced] Pretty-print the handle record including the handle record indices. Hint: with_index=True


(2.2) Decrease server interactions [optional]
---------------------------------------------

The method `get_value_from_handle()` accesses the Handle Server each time. This is a performance slowdown. To avoid it, it is possible to retrieve the record once and then pass it on to the reading methods.

This retrieves the record from the server:

In [11]:
record_json = client.retrieve_handle_record_json(handle)

If you pass it to the reading methods, these do not access the Handle Server anymore:

In [14]:
#print client.pretty_print_handle_record(handle, record_json)

In [15]:
print client.get_value_from_handle(handle, 'CREATION_DATE', record_json)

None


> Task 2.2: Retrieve the handle record as a json object and use the read methods without accessing the Handle Server


(3) Updating handle records
=======================

Now we'd like to modify one of the values of that handle record. The client provides a method for this: `modify_handle_value(handle, ...)`.

Let's try it - let's change the creation date from '2016-05-24' to '2016-05-25':
[Attention - this command will throw an error, this is expected!]

In [7]:
client.modify_handle_value(handle, CREATION_DATE='2016-05-25')

NameError: name 'handle' is not defined

Ups - that did not work. It tells us *'HandleAuthenticationError: Insufficient permission: No credentials passed. Read access only.'*.

(3.1) Write access to a Handle Server
-------------------------------------

For modifying/creating/reading handle records, we first need to authenticate. In this tutorial, we will use a username and a password. (There is other methods, e.g. using client certificates.)

In [1]:
user = 'f'
pw = 'lunchbreak'

Also, we need to specify the URL of a handle server to write to. It is not possible to write to the Global Handle Server. For each prefix, there is only one server with write access.

In [2]:
url = 'https://handle8.dkrz.de'

This server is the write server for the test prefix we are going to use:

In [15]:
prefix = '10876.test'

For providing these necessary configuration values, we can use the following method to create our client object:

In [16]:
client = Client.instantiate_with_username_and_password(url, user, pw, HTTPS_verify=False)

/root/miniconda2/envs/jupytervenv/lib/python2.7/site-packages/requests/packages/urllib3/connectionpool.py:821: InsecureRequestWarning: Unverified HTTPS request is being made. Adding certificate verification is strongly advised. See: https://urllib3.readthedocs.org/en/latest/security.html
  InsecureRequestWarning)


> Task 3.1: Instantiate the client for write access, using the above username and password.

(3.2) Modifying handle records
-------------------------------

### Modify existing values

Now we can try to modify again:

In [17]:
client.modify_handle_value(handle, CREATION_DATE='2016-05-25')

Did it work? Let's check!

In [18]:
print client.get_value_from_handle(handle, 'CREATION_DATE')

2016-05-25


### Adding new values
With the same method, we can add new values to a handle record:

In [19]:
client.modify_handle_value(handle, PLEASE_ADD_ME='added value')
print client.pretty_print_handle_record(handle)

**********************************************************************
* 10876.test/rdapidxyz                                               *
**********************************************************************
* PLEASE_ADD_ME : added value                                        *
* CREATION_DATE : 2016-05-25                                         *
* URL           : http://dkrz.de                                     *
* VERSION_NUMBER: 1                                                  *
* IS_PART_OF    : 10876.test/rdapid44645074-667d-468f-92ca-b73ab ... *
**********************************************************************


To prevent adding new values (for example in case of typos) we can set a flag:

In [20]:
client.modify_handle_value(handle, CREATION_DATE='2016-05-26', add_if_not_exist=False)
client.modify_handle_value(handle, RCEATION_DATE='2016-05-27', add_if_not_exist=False)
print client.pretty_print_handle_record(handle)

**********************************************************************
* 10876.test/rdapidxyz                                               *
**********************************************************************
* CREATION_DATE : 2016-05-26                                         *
* PLEASE_ADD_ME : added value                                        *
* URL           : http://dkrz.de                                     *
* VERSION_NUMBER: 1                                                  *
* IS_PART_OF    : 10876.test/rdapid44645074-667d-468f-92ca-b73ab ... *
**********************************************************************


### Deleting values
In case we did not set that flag and accidentally wrote a wrong entry, the `delete_handle_value()` method allows to delete that entry:

In [21]:
client.modify_handle_value(handle, RCEATION_DATE='2016-05-28')

print 'Added wrong value:'
print client.get_value_from_handle(handle, 'RCEATION_DATE')

client.delete_handle_value(handle, 'RCEATION_DATE')

print 'Deleted wrong value:'
print client.get_value_from_handle(handle, 'RCEATION_DATE')

Added wrong value:
2016-05-28
Deleted wrong value:
None


> Task 3.2a: Modify an existing handle value in your test handle `10876.test/rdapid[xyz]` (with [xyz] being your assigned number).

> Task 3.2b: Attempt to modify handle value that does not exist yet and prevent the library from creating it.

> Task 3.2c: Add a handle value to the handle record.

> Task 3.2d: Add several new handle values to the handle record at the same time (with one command).

> Task 3.2e: Delete one of the handle values previously added.

(4) Registering handles (creating handle records)
=================================================

In their most simple form, PIDs are simple redirection to a URL. In this case, all they have is an entry that stores the URL. You can simply create such a handle using the method `register_handle()`:

In [22]:
newhandle = prefix+'/myhandle'
url = 'www.dkrz.de'
client.register_handle(newhandle, url)

HandleAlreadyExistsException: Handle 10876.test/myhandle already exists: Could not register handle.

If we execute this code a second time - or if another participant of the course has already executed it -, we run into an error: "HandleAlreadyExistsException". Of course - that handle record has already been created!
If we are really sure we would like to overwrite it, we can specify this:

In [23]:
client.register_handle(newhandle, url, overwrite=True)

u'10876.test/myhandle'

To avoid running into this problem and having to decide whether to overwrite or not, it is always preferrable to use UUIDs as handle suffixes. Using "speaking names", i.e. suffixes with semantics, is strongly discouraged.

The library provides an easy way to generate such a handle name. In this case, don't forget to store the handle name in a variable for further use.

In [24]:
uuidhandle = client.generate_PID_name(prefix)
print uuidhandle

10876.test/c8c2ab00-0414-45d5-89d2-af888456d221


And it's even easier to generate the name and register the handle at the same time. Again, don't forget to store the name.

In [25]:
uuidhandle = client.generate_and_register_handle(prefix, url)
print uuidhandle

10876.test/d39f01df-43fd-4b49-a2b4-bfa6910c7927


We can check the contents of this newly created handle record:

In [26]:
print client.pretty_print_handle_record(uuidhandle)

***************************************************
* 10876.test/d39f01df-43fd-4b49-a2b4-bfa6910c7927 *
***************************************************
* URL: www.dkrz.de                                *
***************************************************


> Task 4a: Register a handle with a uuid name and a simple record (containing just a URL).

> Task 4b: Add a value to it.

> Task 4c: Register a handle with a uuid name and several values at the same time.

(5) Authentication - how does it work?
======================================

(5.1) Username and password
----------------------------

Above, we have seen authentication using a username and a password. The username was `300:10876.test/USERLIST`, where `10876.test/USERLIST` is a handle under the `10876.test` prefix, and 300 is an index value to point to a specific entry in this handle.
Let's have a look at this handle:

In [27]:
adminhandle = '10876.test/USERLIST'
adminrecord = client.retrieve_handle_record_json(adminhandle)
print client.pretty_print_handle_record(adminhandle, adminrecord, print_special_types=True, with_index=True)

*********************************************************************
* 10876.test/USERLIST                                               *
*********************************************************************
* 100: HS_ADMIN: {u'index': 201, u'handle': u'10876.TEST/ADMIN' ... *
*********************************************************************


We see no entry with index 300. This is because it's hidden, and this is because it is our password. If it was not hidden, the record would look as follows:

~~~
*********************************************************************
* 10876.test/USERLIST                                               *
*********************************************************************
* 100: HS_ADMIN: {u'index': 201, u'handle': u'10876.TEST/ADMIN' ... *
* 300: HS_VLIST: lunchbreak                                         *
* 301: HS_VLIST: coffeebreak                                        *
*********************************************************************
~~~

The entry with index 300 is of the type "HS_SECKEY" and contains our password.

(5.2) Authentication using client certificates [OPTIONAL]
==============================================

There is an alternative to using password, which is more secure, which is using client-side certificates. For this, the user provides his private key and his certificate with every write request. (Advanced: Creating the key pair and the certificate file in the required format and with the required information is out of scope here. Please check out B2Handle documentation).

### How does it work?

Similar to above, where the username points to "HS_SECKEY", where the password is stored, the username now points to an entry of type "HS_PUBKEY", where the public key is stored. The handle record above contains various of such public keys.

NOTE:
Usernames in the handle systems are strings of the form "index:prefix/suffix" pointing to an entry of type "HS_SECKEY" containing the password of that user, or to an entry "HS_PUBKEY" containing the public key of that user.

### Usage of client-side certificates

To use client-side certificates for authentication, the user has to pass a certificate and a private key along with every write request. Just like the password authentication, the library handles this for the user.

For this, the library either needs a file containing private key and certificate, or both as separate files. To simplify those three different ways of authenticating, there is a special class, called PIDClientCredentials.

In [28]:
from pyhandle.clientcredentials import PIDClientCredentials
cred = PIDClientCredentials(
    handle_server_url='https://handle8.dkrz.de',
    username='301:10876.test/USERLIST',
    private_key='/tmp/dummykeyfile.pem',
    certificate_only='/tmp/dummycertfile.pem'
)

...or, even more simple...

In [29]:
cred = PIDClientCredentials.load_from_JSON('/tmp/credentials_workshop.json')
client = Client.instantiate_with_credentials(cred)

where the path points to a JSON file containing the necessary info, e.g. like this:

~~~
{
    "handle_server_url"="https://handle8.dkrz.de",
    "username": "index:prefix/suffix",
    "private_key"="/path/to/private/key/file.pem",
    "certificate_only"="/path/to/certificate/file.pem"

}
~~~

The other authentication methods can also be passed as JSON, such as:

~~~
{
    "handle_server_url"="https://handle8.dkrz.de",
    "username": "index:prefix/suffix",
    "password": "blablabla"
}
~~~

Or:

~~~
{
    "handle_server_url"="https://handle8.dkrz.de",
    "username": "index:prefix/suffix",
    "certificate_and_key"="/path/to/certificate/and/key/file.pem"
}
~~~

The class PIDClientCredentials reads the JSON file and checks if the necessary information for at least one of these authentication methods are present.

(6) Who may modify my handle record? - Authorisation

So, can I modify and update any handle, provided that I have a username and password (or key pair)? Obviously not.

To modify a handle, two things are necessary:

1. The user needs to be given write permission in general by the owner of the prefix ("Authorisation").
2. The user needs to be given update permissions for the handle record in question.

A little introduction to authorisation is in the next section (but will not be looked into deeply, as this mainly concerns Handle Server administrators).

Update permissions for specific handles are more important, so let's go.

(6) Handle owners
==================

For every handle, it has to be determined who has the right to update or even delete the handle. In other words, every handle has an "owner" (Note that the handle prefix owner always has to rights to change/delete a handle). A handle owner can be one user, or a group of users.

The library by default sets the user group "200:0.NA/prefix" as the handle owner. In general, the entry with index 200 of the prefix owner handle record points to a group of users containing the handle server administrators. This default can be overwritten by adding a "handleowner" to the credentials JSON file. We recommend that users set themselves as the handle owners, or that sensible user groups are defined.

~~~
{
    "handle_server_url"="https://handle8.dkrz.de",
    "username": "300:10876.test/admin",
    "password": "blablabla",
    "handleowner": "300:10876.test/admin"
}
~~~

(Info: User groups are defined using entries of type "HS_VLIST" - for more details on this, see section on Authorisation or the B2Handle documentation).


(7) Advanced: Authorisation [OPTIONAL]
======================================

This section is presented using slides. This is only a very short summary.

For a username and password or client certificate to work, the user needs to be given write permission by the owner of the prefix.

This is done using entries of the type "HS_ADMIN" and "HS_VLIST". The prefix owner can specify a set of permissions in an entry of type "HS_ADMIN", in the prefix owner handle ("0.NA/prefix"):

~~~
{
    "index": 100,
    "type": "HS_ADMIN",
    "data": {
         "format": "admin",
         "value": {
             "handle": "10876.test/admin",
             "index: 300,
             "permissions":"011111110011"
          }
      }
 }
 ~~~

(This gives a specific set of permissions to our user, 300:10876.test/admin.)

Permissions are stored as a string of 12 bits, each of which represents a specific permission that a user/user group can have or not. The prefix owner can grant a set of permissions to one user or to a group of users. To give them to a group of users, he can list these users in an entry of type "HS_VLIST".

An HS_VLIST can again point to a HS_VLIST, possibly creating a cascading chain of users. This allows for a detailed user management. For more details on this, and for a suggestion of Best Practices, please check the online documentation of the B2Handle library (http://eudat-b2safe.github.io/B2HANDLE/givingpermissiontousers.html).

> Task 7: [optional] Retrieve entire record of handle *10876.test/rdapidxyz* as a dictionary, including the HS_ADMIN. Hint: hide_complex_types=False.


Wrap-Up
========

## Using pyhandle

* Instantiate the client for read-only or for write access
* Read-only allows to read the whole record or individual values
* Write-access requires access information (credentials) to a Handle Server (at least 8.1)
* Credentials can be username+password or username+client certificate
* Credentials can be stored in JSON files (use with PIDClientCredentials)


## Recommendations

* Use client certificates for authentication
* Store credentials in JSON file
* Set yourself or a group as handle owner



## Documentation on ReadTheDocs:

Check out B2Handle documentation (and soon pyhandle documentation) for detailed infos!

http://eudat-b2safe.github.io/B2HANDLE

Detailed instructions for:

* Usage of the library
* Authenticati
* Authorisation / User management
* Handle Server administration
* Creation of client certificates and key pairs